### DISCUSSION TITLE
Behavior of getSuccessors()

### DISCUSSION THREAD

**Discussion by alexandernst** [clientIO/joint#1746] [Q&A] [answered]
I made a small demo [0] that is using `getSuccessors()` to find all the elements that are connected to an element.
If I understand the docs correctly, `getSuccessors()` might consider visiting embedded elements if the `deep` opt is set to `true`. 

Note that, however, in this demo `getSuccessors()` is returning 3 children (blue, orange and red in the console). The red element is embedded inside the orange element and it's connected to the "in" port of the orange element.

I'm guessing that `getSuccessors()` doesn't really care if the port are "in" or "out" and just follows any links it can find.
If this is the expected behavior, is there a way I can prevent `getSuccessors()` from following links which's source is an "in" port?
 
[0] - https://codesandbox.io/s/jointjs-dom-graph-forked-fptgmd

**Comment by kumilingus** [comment]
The `deep` option means that the elements connected to any of the ~~ancestors~~  are also successors.

What the `getSuccessors()` care about is the direction of the links. It follows links from their `source` to `target` only.

1. There is no such concept as `in` and `out` ports in JointJS. You have only created 2 groups and named them `in` and `out` (there is no semantic information attached to that).
2. You didn't connect the links to ports in the demo at all.

That means you'll have to write your own function to find all successors (see `graph.dfs`/ `graph.bfs` for inspiration) to incorporate your own logic. But if a port is an output port it should not have any inbound links i.e. it should not be found by the `getSuccessors()`.

---

**EDIT:**
The `deep` option means that the elements connected to any of the descendants are also successors.

**Comment by alexandernst** [comment]
Hi Roman!

> There is no such concept as in and out ports in JointJS

I stand corrected. I mentally built the concepts "in" and "out" ports because I'm used to differentiate them that way in the project I'm working on. 

> You didn't connect the links to ports in the demo at all

Indeed, I posted an incomplete demo. Let me post a fixed demo, just in case somebody in the future finds this useful: https://codesandbox.io/s/jointjs-dom-graph-forked-slj4gt

In the new demo I have applied the following fixes:

* `link` is connected from the "out" port of `rect` (green) to the "in" port of `rect2` (blue)
* `link2` is connected from the "out" port of `rect` (green) to the "in" port of `rect3` (orange)
* `link3` as connected to `rect3` and `rect4` (red; as target)


> The deep option means that the elements connected to any of the ancestors are also successors.

This doesn't match [what the docs state](https://resources.jointjs.com/docs/jointjs/v3.5/joint.html#dia.Graph.prototype.getSuccessors) nor the tests that I have been performing. It doesn't make any sense for ancestors to even be considered in a method called `get*Successors*()`. But it does make sense for `getSuccessors` to find nested (embedded) elements if the `deep` option is set to true.

If I understand you correctly, `getSuccessors()` is finding `rect4` (red) as a successor of `rect` (green) because it's following the links that are connected to the port of `rect3` (orange), but I really think that this shouldn't be happening, as `rect4` is an embedded element and I'm not using the `deep` option.

Can you please double check the documentation of the `getSuccessors()` function and confirm what should `deep` do?

**Reply by kumilingus** [reply]
The documentation for the `deep` option is very unclear. We'll update it.

<img width="518" alt="image" src="https://user-images.githubusercontent.com/3967880/180438258-701fe037-50ad-493c-9528-dc7d1b083f3d.png">

It means that the successors of `A` is only `D` when `deep: false`. And `C` and `D` when `deep: true`.

The current algorithm (`deep: false`) simply follows links. It does not care for embedding. And I think it should stay this way.

<img width="371" alt="image" src="https://user-images.githubusercontent.com/3967880/180440170-0799b479-bd7c-47cb-bf56-e73236799e45.png">

The successor of `A` is currently `B`. In your opinion, the answer should be that there are no successors.

What about this example? I don't think it's that clear here.

<img width="340" alt="image" src="https://user-images.githubusercontent.com/3967880/180440494-58ec44ac-5886-4173-8ea2-8d5dc7eccbe2.png">

The successors of `A` are both `C` and `B`. The result is the elements that I visit if I follow the directed links.

---

I think in your case you should simply call `getSuccessors` and filter all elements that are embedded in any other element from the result afterward.

```js
graph.getSuccessors(el).filter((s1, _, successors) => {
    const isEmbeddedInSuccessor = [el, ...successors].find(s2 => s1 !== s2 && s1.isEmbeddedIn(s2));
    return !isEmbeddedInSuccesor;
});
```

What do you think?

**Comment by alexandernst** [comment, marked as answer]
IMHO, in your first example, `C` and `D` being successors of `A` doesn't sound logical, even if `deep` is set to `true`. The successors of `A` should be any elements that are connected to `A`. If `deep` is set to `true`, nested elements should also be considered. But in no reasonable way the parent of `A` should be considered as a starting point when traversing the graph. Even the semantics of the word `deep` already provide enough evidence that the direction of the traversal should go `down / deep`, not `up / narrow`.

If I had the following graph:

```
┌────────────────────────────────┐
│          A                     │
│                                │
│    ┌───────────────────────┐   │
│    │                       │   │
│    │                       │   │                    ┌───────┐
│    │     ┌──────────────┐  │   │                    │       │
│    │     │              │  │   │                    │   X   │
│    │     │   ┌───────┐  │  │   │                    │       │
│    │     │   │       ├──┼──┼───┼───────────────────►└───────┘
│    │ B   │   │   D   │  │  │   │
│    │     │   │       │  │  │   │
│    │     │   └───────┘  │  │   │
│    │     │              │  │   │
│    │     │      C       │  │   │                   ┌─────────┐
│    │     │              │  │   │                   │         │
│    │     │              │  │   ├──────────────────►│   Y     │
│    │     └──────────────┘  │   │                   │         │
│    │                       │   │                   └─────────┘
│    └───────────────────────┘   │
│                                │
│                                │
└────────────────────────────────┘
```

What would the successors of `D` be (if I set `deep` to `true`)? Are there any limits to how many parents the `getSuccessors()` algorithm will traverse?

> The current algorithm (deep: false) simply follows links. It does not care for embedding. And I think it should stay this way.

This sounds strange, since `getSuccessors()` already cares about embedding (or at least that's what it appears; it's just that it cares about embedding going up, not down).

In your 3rd example, IMHO, the successors of `A` should be `C` if `deep: false` or `C` and `B` if `deep: true`.
Same thing goes for the 2nd example: `A` shouldn't have any successors if `deep: false`.

I guess that the `getSuccessor()` method is very prone to personal interpretation of what `deep / narrow` and `connected` really mean; it's just that I genuinely didn't expect `deep` to mean `up`. I'm not entirely sure if there is a problem with the implementation of `getSuccessors()`, if there is a place for another method (`getGraphSuccessors() maybe?`) or if this should be implemented on the application level and not on the library level.

**Reply by kumilingus** [reply]
> What would the successors of D be (if I set deep to true)? Are there any limits to how many parents the getSuccessors() algorithm will traverse?

It would be only `X`.  Of course, I meant **descendants** and not ancestors. Sorry.

When `deep: true` it cares about embedding going down.

Does this fact change anything in your view?

**Reply by kumilingus** [reply]
What kind of answer in human language are you searching for?

The `getSuccessors` returns elements that you can visit from an element when you follow directed links.
If `deep` is true, an element `A` is considered to be a successor of `B` not only if it is connected to the element `B` but also to any of its descendants.

**Reply by alexandernst** [reply]
Ooohh! If you ment to say `descendants` then some things start to make sense!

Let's go back to the basics and see where do our POVs differ.

## Case 1

This is as simple as it gets. `A` is a successor of `B`. I guess we can both agree on this one.

```
┌────────┐           ┌────────┐
│        │           │        │
│        │           │        │
│  B     ├──────────►│   A    │
│        │           │        │
└────────┘           └────────┘
```

## Case 2

If I understand your explications correctly, `B` doesn't have any successors if `deep: false`. If `deep: true`, then `A` is a successor. Is that correct?

```
┌──────────────────┐
│                  │
│       ┌─────┐    │              ┌───────┐
│       │     │    │              │       │
│       │ C   ├────┼─────────────►│  A    │
│       │     │    │              │       │
│       └─────┘    │              └───────┘
│  B               │
│                  │
└──────────────────┘
```

## Case 3

`C` doesn't have any successor, even if `deep: true`. Is that correct?

```
┌──────────────────┐
│                  │
│       ┌────────┐ │              ┌─────────┐
│       │        │ │              │         │
│       │  C     │ │              │         │
│       │        │ ├─────────────►│   A     │
│       └────────┘ │              │         │
│   B              │              └─────────┘
│                  │
└──────────────────┘
```

## Case 4

If `deep: false`, the successor of `A` is `B`. What about if `deep: true`? Should `C` be marked as a successor? (I think **no**).

```
                         ┌────────────────────────┐
┌─────────┐              │                        │
│         │              │         ┌─────────┐    │
│         │              │         │         │    │
│   A     ├─────────────►│         │         │    │
│         │              │         │   C     │    │
└─────────┘              │         │         │    │
                         │         └─────────┘    │
                         │                        │
                         │   B                    │
                         │                        │
                         └────────────────────────┘
```

## Case 5

What about now? Is `C` a successor of `A` if `deep: false`? IMHO it shouldn't be. But it should If `deep: true`.

```
                         ┌────────────────────────┐
┌─────────┐              │                        │
│         │              │         ┌─────────┐    │
│         │              │         │         │    │
│   A     ├─────────────►│         │         │    │
│         │              │         │   C     │    │
└─────────┘              ├────────►│         │    │
                         │         └─────────┘    │
                         │                        │
                         │   B                    │
                         │                        │
                         └────────────────────────┘
```

**Reply by kumilingus** [reply]
yeah, sorry for the confusion 🤦 😄 

---

**1-3**: correct.

---

**4**: is correct. There is no link between `B` and `C`.

If someone is looking for the result where a successor is any element that is connected directly or through any of its ancestors then this is the solution:

```js
const embedsIncluded = [];
graph.getSuccessors(el).forEach(s1 => {
    const embeds = s1.getEmbeddedCells({ deep: true }).filter(s2 => s2.isElement() && !embedsIncluded.includes(s2));
    embedsIncluded.push(s1, ...embeds);
});
```
---

**5**: the current behavior is IMHO better because it allows both usages (my interpretation and yours).

i.e. I believe that you **CAN** arrive from the current result to the result you are looking for. But it **does not work** the other way round.

Current:
```js
graph.getSuccessors(el)
```

Yours:
```js
graph.getSuccessors(el).filter((s1, _, successors) => {
    const isEmbeddedInSuccessor = [el, ...successors].find(s2 => s1 !== s2 && s1.isEmbeddedIn(s2));
    return !isEmbeddedInSuccesor;
});
```

**Reply by alexandernst** [reply]
Ooohh, I see. So... 🤔... yes, I guess the current behavior is better. Yes, I think now I understand how `getSuccessors()` works and I can implement what I was trying :)

**Reply by kumilingus** [reply]
Here's the ticket to improve the docs: https://github.com/clientIO/joint/issues/1747

Thanks, if you find that I was wrong, feel free to open the discussion again.
